### ISSUE TITLE
MySQL 5.7.9 installation note hidden

### ISSUE THREAD

**Issue by sander3** [Homebrew/homebrew#46114]
Hi,

During the post-install of MySQL 5.7.9 a temporary password is generated for root@localhost. But because Brew doesn't show this temporary password you won't be able to login.

**Comment by renatomoya** [comment]
I'm having the same issue. Just installed MySQL 5.7.9 and I can't log in as root because I don't know where the password is.

**Comment by DomT4** [comment]
Welp. You can get around this for now by removing the existing database in `HOMEBREW_PREFIX/var/mysql` generated by `postinstall` (Providing one didn't already exist - We don't overwrite existing databases), and then run `brew postinstall mysql -v` to see the output.

Will look at something that doesn't require user intervention.

**Comment by sander3** [comment]
cat ~/Library/Logs/Homebrew/mysql/01.mysqld maybe?

**Comment by DomT4** [comment]
If you're installing a bottle that would work. If building from source it'd be a lot further down the log number.

**Comment by renatomoya** [comment]
`cat ~/Library/Logs/Homebrew/mysql/01.mysqld` works.

I re-initialized mysql server like this:

```bash
rm -rf /usr/local/var/mysql
mysqld --initialize-insecure
```

This initializes mysql server with root account with an empty password.

**Comment by sander3** [comment]
Yeah I did that too. But it doesn't solve the problem that the user isn't notified about the user credentials.

This should do the trick:
```
  def post_install
    # Make sure the datadir exists
    datadir.mkpath
    unless (datadir/"mysql/user.frm").exist?
      ENV["TMPDIR"] = nil
      system bin/"mysqld", "--initialize-insecure", "--user=#{ENV["USER"]}",
        "--basedir=#{prefix}", "--datadir=#{datadir}", "--tmpdir=/tmp"
    end
  end

  def caveats; <<-EOS.undent
    A "/etc/my.cnf" from another install may interfere with a Homebrew-built
    server starting up correctly.
    To connect:
        mysql -u root
    Please don't forget to change the root user's password.
    EOS
  end
```

**Comment by DomT4** [comment]
I'd rather not routinely leave people's `mysql` databases unprotected unless there's absolutely no other way to avoid it.

**Comment by sander3** [comment]
You could run ```mysql_secure_installation```.

**Comment by DomT4** [comment]
Discussing a workaround with the team. If nobody on the team objects I'll push it to the formula later today.

**Comment by mikemcquaid** [comment]
I'd probably rather we left things as they were with 5.6 with `--initialize-insecure` and then encourage them to run `mysql_secure_installation` in the caveats themselves. While this may seem insecure given we are OS X and not a Linux package manager I think we should optimise for usability here.

**Comment by xu-cheng** [comment]
> I'd probably rather we left things as they were with 5.6 with --initialize-insecure

I'm strongly uncomfortable to introduce security regression by ignoring upstream's decision. I don't think any package manager is in the position to make such decision. And more certainly, being a OS X package manager shouldn't make any difference regarding security.

> I think we should optimise for usability here.

I don't think initialize insecure database can offer any user experience. In the end, users have to manually delete the database and create one by their own. Therefore, it actually makes UX worse.

IMHO, we should ask user to initial database in caveats. Therefore, I think we should make `mysql.rb` similar like this:
```ruby
def post_install
  datadir.mkpath
end

def caveats
  s = <<-EOS.undent
    A "/etc/my.cnf" from another install may interfere with a Homebrew-built
    server starting up correctly.

  EOS

  s += <<-EOS.undent unless (datadir/"mysql/user.frm").exist?
    To initialize database:
        mysql --initialize --user=$(whoami) --basedir=#{prefix} --datadir=#{datadir} --tmpdir=/tmp

  EOS

  s += <<-EOS.undent
    To connect:
        mysql -uroot
  EOS
end
```

cc @apjanke for thoughts.

**Comment by mikemcquaid** [comment]
> I'm strongly uncomfortable to introduce security regression by ignoring upstream's decision.

I'm uncomfortable with the usability regression. It's not ignoring their decision if we don't patch anything but use a flag they provide and support.

> I don't think initialize insecure database can offer any user experience. In the end, users have to manually delete the database and create one by their own. Therefore, it actually makes UX worse.

Why do they have to do that? What I'm advocating for is the same behaviour as MySQL 5.6. Why did you have no objections to the behaviour then but do now?

> mysql -uroot

Well, they'll need to pass a password now and we don't flag that. You'll also need to enter that password for every application that uses a root MySQL user to do initial privilege setup e.g. a bunch of Rails guides and the like.

**Comment by apjanke** [comment]
I think I'm walking in to an longer-lived discussion here and don't understand the full context, so take anything I say with a large grain of salt.

I think it makes sense to defer to upstream in terms of tightening or loosening security in newer releases even if it breaks some use cases. We're just packagers. At any rate, creating a db with no password, or a pw in plaintext in a well-known location gives me the willies; it feels like a holdover from a single-user system case, and given how connected a "modern" OS X system is, I don't know if that's appropriate.

I think a real big thing to consider in this particular case is that it seems like the MySQL installer is assuming the installer output is ephemeral: that a human will see it, and copy it somewhere else, and then it will disappear. If we're capturing logs, or otherwise storing the output in persistent files, that changes things. I don't know if upstream was considering automated install uses when they decided to do this particular thing.

In summary, my personal view is that creating a no-password or plaintext-password db by default is a bad idea on security grounds, but as a packaging system, Homebrew should just defer to upstream, even if the policy changes between releases. And I'm not an expert here.

**Comment by sander3** [comment]
I'm with @xu-cheng on this.

**Comment by DomT4** [comment]
Feel free to keep the discussion active, but the issue itself was resolved via https://github.com/Homebrew/homebrew/commit/95e8ec6a427f533adcb2f9f5ea5fde881465a1eb.
